src/app/application.cpp

```diff
@@ -754,10 +754,12 @@ try
         });
 
         disconnect(m_desktopIntegration, &DesktopIntegration::activationRequested, this, &Application::createStartupProgressDialog);
-        delete m_desktopIntegration->menu();
+        // we must not delete menu while it is used by DesktopIntegration
```

Comment:
will it simplify things if DesktopIntegration take the ownership of the menu?

Reply:
> will it simplify things if DesktopIntegration take the ownership of the menu?

Due to the strange behavior of QSystemTrayIcon, we still have to delete the old menu after assigning a new one, otherwise I observed (on some platform) the displaying of two menus at the same time.

Reply:
Besides, we can't use Qt parent-child handling, since the QMenu inherits QWidget, so it requires the QWidget as a parent.

Reply:
I have an idea to completely encapsulate the menu handling in DesktopIntegration, so MainWindow could re-use its Actions (in the opposite of current relationship).
(MainWindow still does some unrelated jobs from the old days when qBittorrent was a simple MainWindow-centric application.)
But this is the subject of a separate PR.

Reply:
>I have an idea to completely encapsulate the menu handling in DesktopIntegration, so MainWindow could re-use its Actions (in the opposite of current relationship).
(MainWindow still does some unrelated jobs from the old days when qBittorrent was a simple MainWindow-centric application.)
But this is the subject of a separate PR.

OK, hope it could make it into v4.5.